src/Compilers/CSharp/Portable/Compilation/CSharpDiagnosticFilter.cs

```diff
@@ -133,9 +153,16 @@ public static Diagnostic Filter(Diagnostic d, int warningLevelOption, NullableCo
                 {
                     Syntax.NullableContextState.State.Enabled => true,
                     Syntax.NullableContextState.State.Disabled => false,
-                    _ => nullableOption == NullableContextOptions.Enable || nullableOption == NullableContextOptions.Warnings
+                    Syntax.NullableContextState.State.ExplicitlyRestored => areWarningsGloballyEnabled(),
+                    Syntax.NullableContextState.State.Unknown =>
+                        tree?.IsGeneratedCode(syntaxTreeOptions) != true && areWarningsGloballyEnabled(),
+                    null => areWarningsGloballyEnabled(),
+                    _ => throw ExceptionUtilities.UnexpectedValue(warningsState)
                 };
 
+                bool areWarningsGloballyEnabled()
```

Comment:
I found this very hard to read. Please move it the end of the current scope, and indent the `=>` or put it on the same line as the declaration.